In [1]:
!git clone https://github.com/Narasimhan15/NammaFlow.git

Cloning into 'NammaFlow'...
remote: Enumerating objects: 20, done.
remote: Counting objects: 100% (20/20), done.
remote: Compressing objects: 100% (16/16), done.
remote: Total 20 (delta 6), reused 10 (delta 2), pack-reused 0 (from 0)
Receiving objects: 100% (20/20), 178.77 KiB | 938.00 KiB/s, done.
Resolving deltas: 100% (6/6), done.


In [2]:
%cd /content/NammaFlow

/content/NammaFlow


In [3]:
import os

raw_path = "data/raw"

files = os.listdir(raw_path)

print("Raw datasets:")
for file in files:
    print("-", file)

Raw datasets:
- funnel_live_ward_new_key.json
- city_agg_live_trends_new_key.json
- driver_eda_wards_new_key.json


In [4]:
import json
import os

raw_path = "data/raw"

for filename in files:
    file_path = os.path.join(raw_path, filename)

    with open(file_path, "r") as f:
        data = json.load(f)

    print("\n" + "=" * 80)
    print("FILE:", filename)
    print("Python type:", type(data))

    if isinstance(data, dict):
        print("Top-level keys:", list(data.keys()))

    elif isinstance(data, list):
        print("Number of records:", len(data))

        if len(data) > 0:
            print("First record type:", type(data[0]))

            if isinstance(data[0], dict):
                print("First record keys:", list(data[0].keys()))


FILE: funnel_live_ward_new_key.json
Python type: <class 'list'>
Number of records: 3904
First record type: <class 'dict'>
First record keys: ['ward_num', 'srch_fr_e', 'srch_which_got_e', 'srch_fr_q', 'srch_which_got_q', 'booking', 'ongng_ride', 'done_ride', 'rides', 'cancel_ride', 'drvr_cancel', 'rider_cancel', 'earning', 'dist', 'avg_dist_pr_trip', 'avg_fare', 'trip_category', 'vehicle_type', 'srch_fr_q_variant', 'srch_which_got_q_variant', 'is_bike', 'is_cab', 'is_auto']

FILE: city_agg_live_trends_new_key.json
Python type: <class 'list'>
Number of records: 6947
First record type: <class 'dict'>
First record keys: ['date', 'hour', 'trip_category', 'vehicle_type', 'srch_rqst', 'srch_which_got_e', 'srch_fr_q', 'srch_which_got_q', 'booking', 'rides', 'done_ride', 'earning', 'cancel_ride', 'drvr_cancel', 'rider_cancel', 'srch_fr_q_variant', 'srch_which_got_q_variant', 'cnvr_rate', 'q_accept_rate', 'city', 'booking_type', 'city_code']

FILE: driver_eda_wards_new_key.json
Python type: <cl

In [5]:
import json
import pandas as pd

# File paths
city_file = "data/raw/city_agg_live_trends_new_key.json"
funnel_file = "data/raw/funnel_live_ward_new_key.json"
driver_file = "data/raw/driver_eda_wards_new_key.json"

# Load JSON files
with open(city_file, "r") as f:
    city_data = json.load(f)

with open(funnel_file, "r") as f:
    funnel_data = json.load(f)

with open(driver_file, "r") as f:
    driver_data = json.load(f)

# Convert to DataFrames
city_df = pd.DataFrame(city_data)
funnel_df = pd.DataFrame(funnel_data)
driver_df = pd.DataFrame(driver_data)

print("City dataset shape:", city_df.shape)
print("Funnel dataset shape:", funnel_df.shape)
print("Driver dataset shape:", driver_df.shape)

City dataset shape: (6947, 23)
Funnel dataset shape: (3904, 29)
Driver dataset shape: (486, 6)


In [6]:
print("CITY COLUMNS")
print(city_df.columns.tolist())

print("\nFUNNEL COLUMNS")
print(funnel_df.columns.tolist())

print("\nDRIVER COLUMNS")
print(driver_df.columns.tolist())

CITY COLUMNS
['date', 'hour', 'trip_category', 'vehicle_type', 'srch_rqst', 'srch_which_got_e', 'srch_fr_q', 'srch_which_got_q', 'booking', 'rides', 'done_ride', 'earning', 'cancel_ride', 'drvr_cancel', 'rider_cancel', 'srch_fr_q_variant', 'srch_which_got_q_variant', 'cnvr_rate', 'q_accept_rate', 'city', 'booking_type', 'city_code', 'bkng_cancel_rate']

FUNNEL COLUMNS
['ward_num', 'srch_fr_e', 'srch_which_got_e', 'srch_fr_q', 'srch_which_got_q', 'booking', 'ongng_ride', 'done_ride', 'rides', 'cancel_ride', 'drvr_cancel', 'rider_cancel', 'earning', 'dist', 'avg_dist_pr_trip', 'avg_fare', 'trip_category', 'vehicle_type', 'srch_fr_q_variant', 'srch_which_got_q_variant', 'is_bike', 'is_cab', 'is_auto', 'srch_to_e_rate', 'e_to_q_srch_rate', 'q_accept_rate', 'q_to_bkng_rate', 'cnvr_rate', 'bkng_cancel_rate']

DRIVER COLUMNS
['ward_num', 'active', 'vehicle_type', 'active_drvr', 'drvr_notonride', 'drvr_onride']


In [7]:
for name, df in {
    "CITY": city_df,
    "FUNNEL": funnel_df,
    "DRIVER": driver_df
}.items():

    print("\n" + "=" * 50)
    print(name, "COLUMNS")
    print("=" * 50)

    for i, col in enumerate(df.columns, start=1):
        print(f"{i}. {col}")


CITY COLUMNS
1. date
2. hour
3. trip_category
4. vehicle_type
5. srch_rqst
6. srch_which_got_e
7. srch_fr_q
8. srch_which_got_q
9. booking
10. rides
11. done_ride
12. earning
13. cancel_ride
14. drvr_cancel
15. rider_cancel
16. srch_fr_q_variant
17. srch_which_got_q_variant
18. cnvr_rate
19. q_accept_rate
20. city
21. booking_type
22. city_code
23. bkng_cancel_rate

FUNNEL COLUMNS
1. ward_num
2. srch_fr_e
3. srch_which_got_e
4. srch_fr_q
5. srch_which_got_q
6. booking
7. ongng_ride
8. done_ride
9. rides
10. cancel_ride
11. drvr_cancel
12. rider_cancel
13. earning
14. dist
15. avg_dist_pr_trip
16. avg_fare
17. trip_category
18. vehicle_type
19. srch_fr_q_variant
20. srch_which_got_q_variant
21. is_bike
22. is_cab
23. is_auto
24. srch_to_e_rate
25. e_to_q_srch_rate
26. q_accept_rate
27. q_to_bkng_rate
28. cnvr_rate
29. bkng_cancel_rate

DRIVER COLUMNS
1. ward_num
2. active
3. vehicle_type
4. active_drvr
5. drvr_notonride
6. drvr_onride


In [8]:
# ==========================================
# NAMMAFLOW - BASIC DATA UNDERSTANDING
# ==========================================

# 1. First few rows
print("=" * 70)
print("CITY DATASET - FIRST 5 ROWS")
print("=" * 70)
display(city_df.head())

print("\n" + "=" * 70)
print("FUNNEL DATASET - FIRST 5 ROWS")
print("=" * 70)
display(funnel_df.head())

print("\n" + "=" * 70)
print("DRIVER DATASET - FIRST 5 ROWS")
print("=" * 70)
display(driver_df.head())


# 2. Data types
print("\nCITY DATA TYPES")
print(city_df.dtypes)

print("\nFUNNEL DATA TYPES")
print(funnel_df.dtypes)

print("\nDRIVER DATA TYPES")
print(driver_df.dtypes)


# 3. Date coverage of city dataset
print("\n" + "=" * 70)
print("CITY DATE INFORMATION")
print("=" * 70)

print("Number of unique dates:", city_df["date"].nunique())
print("Minimum date:", city_df["date"].min())
print("Maximum date:", city_df["date"].max())

print("\nUnique hours:")
print(sorted(city_df["hour"].dropna().unique()))


# 4. Important categorical values
print("\nCITY VEHICLE TYPES")
print(city_df["vehicle_type"].value_counts(dropna=False))

print("\nCITY TRIP CATEGORIES")
print(city_df["trip_category"].value_counts(dropna=False))

print("\nCITY BOOKING TYPES")
print(city_df["booking_type"].value_counts(dropna=False))

print("\nFUNNEL VEHICLE TYPES")
print(funnel_df["vehicle_type"].value_counts(dropna=False))

print("\nDRIVER VEHICLE TYPES")
print(driver_df["vehicle_type"].value_counts(dropna=False))


# 5. Ward information
print("\n" + "=" * 70)
print("WARD INFORMATION")
print("=" * 70)

print("Funnel unique wards:", funnel_df["ward_num"].nunique())
print("Driver unique wards:", driver_df["ward_num"].nunique())

CITY DATASET - FIRST 5 ROWS


,date,hour,trip_category,vehicle_type,srch_rqst,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,rides,...,drvr_cancel,rider_cancel,srch_fr_q_variant,srch_which_got_q_variant,cnvr_rate,q_accept_rate,city,booking_type,city_code,bkng_cancel_rate
0,2026-05-03,0,InterCity,All,18,18,18,18,0,0,...,0,0,0.0,0.0,0.0,1.0000,Bengaluru,Normal,BLR,NaN
1,2026-05-03,0,InterCity,Cab,0,0,0,0,3,2,...,0,2,18.0,17.0,NaN,NaN,Bengaluru,Normal,BLR,0.6667
2,2026-05-03,0,OneWay,All,5132,5118,3781,2390,0,0,...,0,0,0.0,0.0,0.0,0.6321,Bengaluru,Normal,BLR,NaN
3,2026-05-03,0,OneWay,Auto,0,0,0,0,1164,1164,...,264,237,1914.0,1180.0,NaN,NaN,Bengaluru,Normal,BLR,0.4304
4,2026-05-03,0,OneWay,Cab,0,0,0,0,936,720,...,134,378,1930.0,1211.0,NaN,NaN,Bengaluru,Normal,BLR,0.5598



FUNNEL DATASET - FIRST 5 ROWS


,ward_num,srch_fr_e,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,ongng_ride,done_ride,rides,cancel_ride,...,srch_which_got_q_variant,is_bike,is_cab,is_auto,srch_to_e_rate,e_to_q_srch_rate,q_accept_rate,q_to_bkng_rate,cnvr_rate,bkng_cancel_rate
0,b_1,0,0,0,0,0,0,0,0,0,...,0,False,False,False,NaN,NaN,NaN,NaN,NaN,NaN
1,b_1,0,0,0,0,0,0,0,0,0,...,0,False,False,True,NaN,NaN,NaN,NaN,NaN,NaN
2,b_1,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
3,b_1,0,0,0,0,0,0,0,0,0,...,0,False,True,False,NaN,NaN,NaN,NaN,NaN,NaN
4,b_1,0,0,0,0,0,0,0,0,0,...,0,False,False,False,NaN,NaN,NaN,NaN,NaN,NaN



DRIVER DATASET - FIRST 5 ROWS


,ward_num,active,vehicle_type,active_drvr,drvr_notonride,drvr_onride
0,b_1,TRUE,Auto,181,172,9
1,b_1,TRUE,Cab,155,141,14
2,b_2,TRUE,Auto,61,58,3
3,b_2,TRUE,Cab,39,37,2
4,b_3,TRUE,Auto,97,91,6



CITY DATA TYPES
date                         object
hour                          int64
trip_category                object
vehicle_type                 object
srch_rqst                     int64
srch_which_got_e              int64
srch_fr_q                     int64
srch_which_got_q              int64
booking                       int64
rides                         int64
done_ride                     int64
earning                       int64
cancel_ride                   int64
drvr_cancel                   int64
rider_cancel                  int64
srch_fr_q_variant           float64
srch_which_got_q_variant    float64
cnvr_rate                   float64
q_accept_rate               float64
city                         object
booking_type                 object
city_code                    object
bkng_cancel_rate            float64
dtype: object

FUNNEL DATA TYPES
ward_num                     object
srch_fr_e                     int64
srch_which_got_e              int64
srch_fr_q     

We are now trying to answer four questions:

**1. Do we genuinely have historical/time-series data?**

date + hour will determine whether we can build hourly demand forecasting.

**2. What exactly are the vehicle categories?**

For example Bike / Auto / Cab or other categories.

**3. How many Bangalore wards do we cover?**

This determines whether we can build the geographic supply-demand component.

**4. What datatype problems exist?**

For example, date may currently be stored as text rather than datetime.

In [9]:
# ============================================================
# NAMMAFLOW - DATA COVERAGE & QUALITY CHECK
# ============================================================

import pandas as pd

# Convert a COPY of the date column for inspection
city_df["date"] = pd.to_datetime(city_df["date"])

print("=" * 70)
print("1. EXACT DATES AVAILABLE")
print("=" * 70)

print(sorted(city_df["date"].dt.strftime("%Y-%m-%d").unique()))


print("\n" + "=" * 70)
print("2. ROWS PER DATE")
print("=" * 70)

print(city_df.groupby("date").size())


print("\n" + "=" * 70)
print("3. NUMBER OF HOURS AVAILABLE PER DATE")
print("=" * 70)

print(city_df.groupby("date")["hour"].nunique())


print("\n" + "=" * 70)
print("4. HOUR RANGE")
print("=" * 70)

print("Minimum hour:", city_df["hour"].min())
print("Maximum hour:", city_df["hour"].max())
print("Unique hours:", sorted(city_df["hour"].unique()))


print("\n" + "=" * 70)
print("5. MISSING VALUES")
print("=" * 70)

print("\nCITY:")
print(city_df.isnull().sum()[city_df.isnull().sum() > 0])

print("\nFUNNEL:")
print(funnel_df.isnull().sum()[funnel_df.isnull().sum() > 0])

print("\nDRIVER:")
print(driver_df.isnull().sum()[driver_df.isnull().sum() > 0])


print("\n" + "=" * 70)
print("6. EXACT DUPLICATE ROWS")
print("=" * 70)

print("City duplicates:", city_df.duplicated().sum())
print("Funnel duplicates:", funnel_df.duplicated().sum())
print("Driver duplicates:", driver_df.duplicated().sum())


print("\n" + "=" * 70)
print("7. FUNNEL TRIP CATEGORIES")
print("=" * 70)

print(funnel_df["trip_category"].value_counts(dropna=False))


print("\n" + "=" * 70)
print("8. FUNNEL VEHICLE × TRIP CATEGORY COMBINATIONS")
print("=" * 70)

print(pd.crosstab(
    funnel_df["vehicle_type"],
    funnel_df["trip_category"]
))


print("\n" + "=" * 70)
print("9. DRIVER COVERAGE BY VEHICLE")
print("=" * 70)

print(
    driver_df.groupby("vehicle_type")["ward_num"]
             .nunique()
             .sort_values(ascending=False)
)


print("\n" + "=" * 70)
print("10. CAB WARDS MISSING FROM DRIVER DATA")
print("=" * 70)

all_wards = set(driver_df["ward_num"].unique())

cab_wards = set(
    driver_df.loc[
        driver_df["vehicle_type"] == "Cab",
        "ward_num"
    ]
)

print(sorted(all_wards - cab_wards))

1. EXACT DATES AVAILABLE
['2026-01-14', '2026-01-20', '2026-01-21', '2026-05-03', '2026-05-09', '2026-05-10']

2. ROWS PER DATE
date
2026-01-14      24
2026-01-20      24
2026-01-21       5
2026-05-03    2369
2026-05-09    2363
2026-05-10    2162
dtype: int64

3. NUMBER OF HOURS AVAILABLE PER DATE
date
2026-01-14    24
2026-01-20    24
2026-01-21     5
2026-05-03    24
2026-05-09    24
2026-05-10    24
Name: hour, dtype: int64

4. HOUR RANGE
Minimum hour: 0
Maximum hour: 23
Unique hours: [np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(20), np.int64(21), np.int64(22), np.int64(23)]

5. MISSING VALUES

CITY:
srch_fr_q_variant            539
srch_which_got_q_variant     539
cnvr_rate                   4150
q_accept_rate               4353
bkng_cancel_rate      

--> Why we're doing this <--

This one cell answers the remaining questions before we clean:

**Do all six dates have 24 hours?**

If yes, the temporal data is structured; if not, we need to detect incomplete snapshots.

**Are there missing values?**

We need this before deciding whether to drop, impute or preserve anything.

**Are there duplicate records?**

We don't automatically remove them until we know what a valid unique row represents.

**What is the ward × vehicle × trip structure?**

I suspect the 3,904 funnel rows follow something close to:

244 wards
× 4 vehicle categories
× 4 trip categories
≈ 3,904 rows

#what I have inspected so far :

**we’ve already learned four important things:**

- There are 6 dates in city_df: Jan 14, Jan 20, Jan 21, May 3, May 9, May 10.

- Jan 21 is incomplete — it has only 5 hours, while the other dates have 24.

- funnel_df has 269 exact duplicate rows. We need to understand why before deleting them.

- driver_df has no missing values or duplicates, but Cab is missing wards b_156 and b_77 compared with the expected ward coverage.

#Next

**For NammaFlow:** determine what ONE ROW represents in each dataset. This is called the **data grain**. Once we know that, we'll know whether those 269 funnel duplicates are genuinely bad data and what columns form the unique key.

In [10]:
# ================================================================
# NAMMAFLOW — IDENTIFY DATA GRAIN / UNIQUE KEYS
# ================================================================

print("=" * 70)
print("1. CITY DATA — UNIQUE VALUES")
print("=" * 70)

city_cols = [
    "date",
    "hour",
    "trip_category",
    "vehicle_type",
    "city",
    "booking_type",
    "city_code"
]

for col in city_cols:
    if col in city_df.columns:
        print(f"\n{col}: {city_df[col].nunique()} unique values")
        print(city_df[col].value_counts(dropna=False).head(15))


print("\n" + "=" * 70)
print("2. FUNNEL DATA — UNIQUE VALUES")
print("=" * 70)

funnel_cols = [
    "ward_num",
    "trip_category",
    "vehicle_type"
]

for col in funnel_cols:
    if col in funnel_df.columns:
        print(f"\n{col}: {funnel_df[col].nunique()} unique values")
        print(funnel_df[col].value_counts(dropna=False).head(15))


print("\n" + "=" * 70)
print("3. DRIVER DATA — UNIQUE VALUES")
print("=" * 70)

driver_cols = [
    "ward_num",
    "vehicle_type"
]

for col in driver_cols:
    if col in driver_df.columns:
        print(f"\n{col}: {driver_df[col].nunique()} unique values")
        print(driver_df[col].value_counts(dropna=False).head(15))


print("\n" + "=" * 70)
print("4. POSSIBLE CITY UNIQUE KEY CHECK")
print("=" * 70)

city_key = [
    "date",
    "hour",
    "trip_category",
    "vehicle_type",
    "booking_type"
]

city_key = [col for col in city_key if col in city_df.columns]

print("Key used:", city_key)
print(
    "Duplicate rows using this key:",
    city_df.duplicated(subset=city_key).sum()
)


print("\n" + "=" * 70)
print("5. POSSIBLE FUNNEL UNIQUE KEY CHECK")
print("=" * 70)

funnel_key = [
    "ward_num",
    "trip_category",
    "vehicle_type"
]

funnel_key = [col for col in funnel_key if col in funnel_df.columns]

print("Key used:", funnel_key)
print(
    "Duplicate rows using this key:",
    funnel_df.duplicated(subset=funnel_key).sum()
)


print("\n" + "=" * 70)
print("6. POSSIBLE DRIVER UNIQUE KEY CHECK")
print("=" * 70)

driver_key = [
    "ward_num",
    "vehicle_type"
]

driver_key = [col for col in driver_key if col in driver_df.columns]

print("Key used:", driver_key)
print(
    "Duplicate rows using this key:",
    driver_df.duplicated(subset=driver_key).sum()
)

1. CITY DATA — UNIQUE VALUES

date: 6 unique values
date
2026-05-03    2369
2026-05-09    2363
2026-05-10    2162
2026-01-14      24
2026-01-20      24
2026-01-21       5
Name: count, dtype: int64

hour: 24 unique values
hour
10    333
12    328
15    326
11    321
8     317
13    317
7     315
9     315
16    309
14    307
18    299
17    299
20    299
19    293
21    292
Name: count, dtype: int64

trip_category: 4 unique values
trip_category
OneWay       2909
InterCity    1897
Rental       1602
All           539
Name: count, dtype: int64

vehicle_type: 4 unique values
vehicle_type
All     2871
Cab     2063
Auto    1581
Bike     432
Name: count, dtype: int64

city: 19 unique values
city
Cuttack               864
Bhubaneshwar          864
Bengaluru             811
Gulbarga              745
Tumkuru               739
Mysore                523
Rest of Karnataka     471
Chennai               453
Thiruvananthapuram    443
Coimbatore            277
Trichy                218
Durgapur         

-->What this cell is doing?

We're asking:
What exactly does one row mean?

**For example, city_df might ultimately represent something like:**
Date × Hour × Trip Category × Vehicle Type × Booking Type

**while funnel_df may represent:**
Ward × Trip Category × Vehicle Type

and **driver_df may represent:**
Ward × Vehicle Type

In [11]:
# ================================================================
# NAMMAFLOW — VERIFY TRUE DATA GRAIN
# ================================================================

print("=" * 75)
print("1. FULL CITY COLUMN LIST")
print("=" * 75)
print(city_df.columns.tolist())


print("\n" + "=" * 75)
print("2. FULL FUNNEL COLUMN LIST")
print("=" * 75)
print(funnel_df.columns.tolist())


# ------------------------------------------------
# CITY: include city in candidate key
# ------------------------------------------------

print("\n" + "=" * 75)
print("3. CITY KEY — INCLUDING CITY")
print("=" * 75)

city_key_v2 = [
    "date",
    "hour",
    "city",
    "trip_category",
    "vehicle_type",
    "booking_type"
]

city_key_v2 = [c for c in city_key_v2 if c in city_df.columns]

print("Key:", city_key_v2)
print(
    "Duplicate rows using this key:",
    city_df.duplicated(subset=city_key_v2).sum()
)


# ------------------------------------------------
# FUNNEL: inspect booking_type
# ------------------------------------------------

print("\n" + "=" * 75)
print("4. FUNNEL BOOKING TYPE CHECK")
print("=" * 75)

if "booking_type" in funnel_df.columns:

    print("\nBooking types:")
    print(funnel_df["booking_type"].value_counts(dropna=False))

    print("\nTrip Category × Booking Type:")
    print(
        pd.crosstab(
            funnel_df["trip_category"],
            funnel_df["booking_type"],
            margins=True
        )
    )

else:
    print("booking_type is NOT present in funnel_df")


# ------------------------------------------------
# FUNNEL candidate key with booking_type
# ------------------------------------------------

print("\n" + "=" * 75)
print("5. FUNNEL KEY — INCLUDING BOOKING TYPE")
print("=" * 75)

funnel_key_v2 = [
    "ward_num",
    "trip_category",
    "vehicle_type",
    "booking_type"
]

funnel_key_v2 = [c for c in funnel_key_v2 if c in funnel_df.columns]

print("Key:", funnel_key_v2)

print(
    "Duplicate rows using this key:",
    funnel_df.duplicated(subset=funnel_key_v2).sum()
)


# ------------------------------------------------
# Exact duplicates
# ------------------------------------------------

print("\n" + "=" * 75)
print("6. EXACT FUNNEL DUPLICATES")
print("=" * 75)

print("Exact duplicate rows:", funnel_df.duplicated().sum())

exact_dup_rows = funnel_df[
    funnel_df.duplicated(keep=False)
].copy()

print("Rows involved in duplicate groups:", len(exact_dup_rows))


# ------------------------------------------------
# Show sample duplicate records
# ------------------------------------------------

print("\n" + "=" * 75)
print("7. SAMPLE FUNNEL DUPLICATE RECORDS")
print("=" * 75)

display(
    exact_dup_rows
    .sort_values(
        [c for c in
         ["ward_num", "trip_category", "vehicle_type", "booking_type"]
         if c in exact_dup_rows.columns]
    )
    .head(20)
)

1. FULL CITY COLUMN LIST
['date', 'hour', 'trip_category', 'vehicle_type', 'srch_rqst', 'srch_which_got_e', 'srch_fr_q', 'srch_which_got_q', 'booking', 'rides', 'done_ride', 'earning', 'cancel_ride', 'drvr_cancel', 'rider_cancel', 'srch_fr_q_variant', 'srch_which_got_q_variant', 'cnvr_rate', 'q_accept_rate', 'city', 'booking_type', 'city_code', 'bkng_cancel_rate']

2. FULL FUNNEL COLUMN LIST
['ward_num', 'srch_fr_e', 'srch_which_got_e', 'srch_fr_q', 'srch_which_got_q', 'booking', 'ongng_ride', 'done_ride', 'rides', 'cancel_ride', 'drvr_cancel', 'rider_cancel', 'earning', 'dist', 'avg_dist_pr_trip', 'avg_fare', 'trip_category', 'vehicle_type', 'srch_fr_q_variant', 'srch_which_got_q_variant', 'is_bike', 'is_cab', 'is_auto', 'srch_to_e_rate', 'e_to_q_srch_rate', 'q_accept_rate', 'q_to_bkng_rate', 'cnvr_rate', 'bkng_cancel_rate']

3. CITY KEY — INCLUDING CITY
Key: ['date', 'hour', 'city', 'trip_category', 'vehicle_type', 'booking_type']
Duplicate rows using this key: 0

4. FUNNEL BOOKING T

,ward_num,srch_fr_e,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,ongng_ride,done_ride,rides,cancel_ride,...,srch_which_got_q_variant,is_bike,is_cab,is_auto,srch_to_e_rate,e_to_q_srch_rate,q_accept_rate,q_to_bkng_rate,cnvr_rate,bkng_cancel_rate
2,b_1,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
10,b_1,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
18,b_10,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
26,b_10,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
34,b_100,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
42,b_100,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
50,b_101,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
58,b_101,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
66,b_102,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN
74,b_102,0,0,0,0,0,0,0,0,0,...,0,True,False,False,NaN,NaN,NaN,NaN,NaN,NaN


#The datasets has revealed few real issues

CITY KEY:
date + hour + city + trip_category + vehicle_type + booking_type

Duplicate rows using this key: 0

FUNNEL_DF: booking_type is NOT present,
ward_num + trip_category + vehicle_type
Duplicate rows = 976, Exact duplicates = 269
Rows involved = 538 and 244 wards × 16 rows per ward = 3,904 rows.

Therefore vehicle categories:
-> All
-> Auto
-> Bike
-> Cab

so, 16 / 4 vehicle types = 4 trip-category positions per ward but the data currently reports only
-> OneWay     1952
-> InterCity   976
-> Rental      976
and Notice : 1952 = 976 × 2

**This strongly suggests the funnel dataset is structurally designed for four trip-category groups, but two groups are currently labelled OneWay.**

**So those 269 "exact duplicates" might not actually be garbage duplicate records. They may belong to two different intended categories whose labels became identical. Deleting them now could destroy valid observations.**

#Next : Identify the mysterious fourth category

In [12]:
# ================================================================
# NAMMAFLOW — INVESTIGATE FUNNEL TRIP-CATEGORY STRUCTURE
# ================================================================

print("=" * 75)
print("1. ROWS PER WARD")
print("=" * 75)

print(
    funnel_df.groupby("ward_num")
    .size()
    .value_counts()
    .sort_index()
)


print("\n" + "=" * 75)
print("2. NUMBER OF RECORDS PER FUNNEL KEY")
print("=" * 75)

funnel_key_counts = (
    funnel_df
    .groupby(["ward_num", "trip_category", "vehicle_type"])
    .size()
)

print(funnel_key_counts.value_counts().sort_index())


print("\n" + "=" * 75)
print("3. KEY MULTIPLICITY BY TRIP CATEGORY")
print("=" * 75)

key_check = (
    funnel_df
    .groupby(["ward_num", "trip_category", "vehicle_type"])
    .size()
    .reset_index(name="records")
)

print(
    key_check
    .groupby(["trip_category", "records"])
    .size()
)


# ------------------------------------------------
# Inspect one complete ward in ORIGINAL row order
# ------------------------------------------------

print("\n" + "=" * 75)
print("4. SAMPLE WARD — ORIGINAL ROW STRUCTURE")
print("=" * 75)

sample_ward = funnel_df["ward_num"].iloc[0]

print("Sample ward:", sample_ward)

sample_cols = [
    "ward_num",
    "trip_category",
    "vehicle_type",
    "srch_fr_e",
    "srch_which_got_e",
    "srch_fr_q",
    "srch_which_got_q",
    "booking",
    "done_ride",
    "rides",
    "cancel_ride"
]

sample_cols = [c for c in sample_cols if c in funnel_df.columns]

display(
    funnel_df.loc[
        funnel_df["ward_num"] == sample_ward,
        sample_cols
    ]
)


# ------------------------------------------------
# Give the two OneWay records occurrence numbers
# ------------------------------------------------

print("\n" + "=" * 75)
print("5. ONEWAY OCCURRENCE CHECK")
print("=" * 75)

oneway = funnel_df[
    funnel_df["trip_category"] == "OneWay"
].copy()

oneway["oneway_occurrence"] = (
    oneway
    .groupby(["ward_num", "vehicle_type"])
    .cumcount()
)

print(
    oneway["oneway_occurrence"]
    .value_counts()
    .sort_index()
)


print("\n" + "=" * 75)
print("6. SAMPLE ONEWAY PAIRS")
print("=" * 75)

display(
    oneway[
        oneway["ward_num"] == sample_ward
    ][
        sample_cols + ["oneway_occurrence"]
    ]
)

1. ROWS PER WARD
16    244
Name: count, dtype: int64

2. NUMBER OF RECORDS PER FUNNEL KEY
1    1952
2     976
Name: count, dtype: int64

3. KEY MULTIPLICITY BY TRIP CATEGORY
trip_category  records
InterCity      1          976
OneWay         2          976
Rental         1          976
dtype: int64

4. SAMPLE WARD — ORIGINAL ROW STRUCTURE
Sample ward: b_1


,ward_num,trip_category,vehicle_type,srch_fr_e,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,done_ride,rides,cancel_ride
0,b_1,OneWay,All,0,0,0,0,0,0,0,0
1,b_1,OneWay,Auto,0,0,0,0,0,0,0,0
2,b_1,OneWay,Bike,0,0,0,0,0,0,0,0
3,b_1,OneWay,Cab,0,0,0,0,0,0,0,0
4,b_1,InterCity,All,0,0,0,0,0,0,0,0
5,b_1,InterCity,Auto,0,0,0,0,0,0,0,0
6,b_1,InterCity,Bike,0,0,0,0,0,0,0,0
7,b_1,InterCity,Cab,0,0,0,0,0,0,0,0
8,b_1,OneWay,All,70,70,37,19,0,0,0,0
9,b_1,OneWay,Auto,0,0,0,0,10,7,10,2



5. ONEWAY OCCURRENCE CHECK
oneway_occurrence
0    976
1    976
Name: count, dtype: int64

6. SAMPLE ONEWAY PAIRS


,ward_num,trip_category,vehicle_type,srch_fr_e,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,done_ride,rides,cancel_ride,oneway_occurrence
0,b_1,OneWay,All,0,0,0,0,0,0,0,0,0
1,b_1,OneWay,Auto,0,0,0,0,0,0,0,0,0
2,b_1,OneWay,Bike,0,0,0,0,0,0,0,0,0
3,b_1,OneWay,Cab,0,0,0,0,0,0,0,0,0
8,b_1,OneWay,All,70,70,37,19,0,0,0,0,1
9,b_1,OneWay,Auto,0,0,0,0,10,7,10,2,1
10,b_1,OneWay,Bike,0,0,0,0,0,0,0,0,1
11,b_1,OneWay,Cab,0,0,0,0,9,5,9,4,1


Every ward has exactly 16 rows : 244 wards × 16 = 3904 rows

For every **ward_num** + **vwhicle_type** :
Slot 1 → labelled OneWay,
Slot 2 → InterCity,
Slot 3 → labelled OneWay,
Slot 4 → Rental

InterCity → 1 record per ward × vehicle
OneWay   → 2 records per ward × vehicle
Rental   → 1 record per ward × vehicle

The real important part is sample b_1: First OneWay block:
All / Auto / Bike / Cab
→ mostly/all zeros

Second OneWay block:
All → searches = 70
Auto → bookings = 10
Cab → bookings = 9


The two **one-way** rows are same wards and vehicles are often different observations : 269 exact duplicates

#Next : Inspect the raw JSON directly

In [13]:
# ================================================================
# NAMMAFLOW — INSPECT RAW FUNNEL JSON STRUCTURE
# ================================================================

import json

funnel_path = "data/raw/funnel_live_ward_new_key.json"

with open(funnel_path, "r") as f:
    funnel_raw = json.load(f)

print("=" * 75)
print("1. RAW JSON TYPE")
print("=" * 75)
print(type(funnel_raw))


print("\n" + "=" * 75)
print("2. TOP-LEVEL STRUCTURE")
print("=" * 75)

if isinstance(funnel_raw, dict):
    print("Top-level keys:")
    print(list(funnel_raw.keys()))

elif isinstance(funnel_raw, list):
    print("Number of top-level items:", len(funnel_raw))


print("\n" + "=" * 75)
print("3. FIRST RAW OBJECT")
print("=" * 75)

if isinstance(funnel_raw, list):
    print(json.dumps(funnel_raw[0], indent=2)[:10000])

elif isinstance(funnel_raw, dict):
    first_key = list(funnel_raw.keys())[0]

    print("First key:", first_key)
    print(
        json.dumps(
            funnel_raw[first_key],
            indent=2
        )[:10000]
    )

1. RAW JSON TYPE
<class 'list'>

2. TOP-LEVEL STRUCTURE
Number of top-level items: 3904

3. FIRST RAW OBJECT
{
  "ward_num": "b_1",
  "srch_fr_e": 0,
  "srch_which_got_e": 0,
  "srch_fr_q": 0,
  "srch_which_got_q": 0,
  "booking": 0,
  "ongng_ride": 0,
  "done_ride": 0,
  "rides": 0,
  "cancel_ride": 0,
  "drvr_cancel": 0,
  "rider_cancel": 0,
  "earning": 0,
  "dist": 0,
  "avg_dist_pr_trip": 0,
  "avg_fare": 0,
  "trip_category": "OneWay",
  "vehicle_type": "All",
  "srch_fr_q_variant": 0,
  "srch_which_got_q_variant": 0,
  "is_bike": "False",
  "is_cab": "False",
  "is_auto": "False"
}


So this is not a mistake caused by Pandas or by how we converted the JSON into a DataFrame.

The source file itself contains two different records under the same OneWay category for every ward_num × vehicle_type.

Also, the JSON is completely flat. There is no hidden parent key like service_type, ride_mode, etc. that we can simply recover.

Before deciding what to do with those two OneWay blocks, we need one final investigation:

**Q. Are the first OneWay records basically empty placeholder rows, or do they also contain genuine activity?**

In [14]:
# ================================================================
# NAMMAFLOW — COMPARE THE TWO ONEWAY BLOCKS
# ================================================================

# Create a temporary copy — raw data remains untouched
ow = funnel_df[
    funnel_df["trip_category"] == "OneWay"
].copy()

# Identify first and second OneWay occurrence
# for each Ward × Vehicle combination
ow["oneway_occurrence"] = (
    ow.groupby(["ward_num", "vehicle_type"])
      .cumcount()
)

# Core operational metrics
metric_cols = [
    "srch_fr_e",
    "srch_which_got_e",
    "srch_fr_q",
    "srch_which_got_q",
    "booking",
    "ongng_ride",
    "done_ride",
    "rides",
    "cancel_ride",
    "drvr_cancel",
    "rider_cancel",
    "earning",
    "dist"
]

metric_cols = [c for c in metric_cols if c in ow.columns]


print("=" * 75)
print("1. ROW COUNT PER ONEWAY OCCURRENCE")
print("=" * 75)

print(
    ow["oneway_occurrence"]
    .value_counts()
    .sort_index()
)


# ------------------------------------------------
# Number of completely zero rows
# ------------------------------------------------

ow["all_core_metrics_zero"] = (
    ow[metric_cols].fillna(0).eq(0).all(axis=1)
)

print("\n" + "=" * 75)
print("2. ZERO vs NON-ZERO ROWS")
print("=" * 75)

print(
    pd.crosstab(
        ow["oneway_occurrence"],
        ow["all_core_metrics_zero"],
        margins=True
    )
)


# ------------------------------------------------
# Aggregate operational activity
# ------------------------------------------------

print("\n" + "=" * 75)
print("3. TOTAL ACTIVITY BY ONEWAY OCCURRENCE")
print("=" * 75)

activity_summary = (
    ow.groupby("oneway_occurrence")[metric_cols]
      .sum()
      .T
)

display(activity_summary)


# ------------------------------------------------
# Count exact equality between the pair
# ------------------------------------------------

print("\n" + "=" * 75)
print("4. EXACTLY IDENTICAL ONEWAY PAIRS")
print("=" * 75)

compare_cols = [
    c for c in funnel_df.columns
    if c not in ["ward_num", "trip_category", "vehicle_type"]
]

ow0 = (
    ow[ow["oneway_occurrence"] == 0]
    .set_index(["ward_num", "vehicle_type"])
    [compare_cols]
    .sort_index()
)

ow1 = (
    ow[ow["oneway_occurrence"] == 1]
    .set_index(["ward_num", "vehicle_type"])
    [compare_cols]
    .sort_index()
)

common_index = ow0.index.intersection(ow1.index)

same_pair = (
    ow0.loc[common_index]
    .reset_index(drop=True)
    .eq(
        ow1.loc[common_index]
        .reset_index(drop=True)
    )
    .all(axis=1)
)

print("Total OneWay pairs:", len(common_index))
print("Exactly identical pairs:", same_pair.sum())
print("Different pairs:", (~same_pair).sum())


# ------------------------------------------------
# Are there cases where BOTH blocks have activity?
# ------------------------------------------------

print("\n" + "=" * 75)
print("5. BOTH ONEWAY RECORDS HAVE ACTIVITY?")
print("=" * 75)

activity_flag = (
    ow.assign(
        has_activity=~ow["all_core_metrics_zero"]
    )
    .pivot_table(
        index=["ward_num", "vehicle_type"],
        columns="oneway_occurrence",
        values="has_activity",
        aggfunc="first"
    )
)

both_active = activity_flag[
    (activity_flag.get(0, False) == True) &
    (activity_flag.get(1, False) == True)
]

print(
    "Ward × vehicle pairs where BOTH OneWay records have activity:",
    len(both_active)
)

display(both_active.head(20))


# ------------------------------------------------
# Show examples where the two blocks differ
# ------------------------------------------------

print("\n" + "=" * 75)
print("6. SAMPLE DIFFERENT ONEWAY PAIRS")
print("=" * 75)

different_keys = common_index[~same_pair.values]

if len(different_keys) > 0:

    sample_keys = list(different_keys[:5])

    display(
        ow[
            ow.set_index(
                ["ward_num", "vehicle_type"]
            ).index.isin(sample_keys)
        ][
            ["ward_num",
             "vehicle_type",
             "oneway_occurrence"] + metric_cols
        ].sort_values(
            ["ward_num",
             "vehicle_type",
             "oneway_occurrence"]
        )
    )

else:
    print("No different pairs found.")

1. ROW COUNT PER ONEWAY OCCURRENCE
oneway_occurrence
0    976
1    976
Name: count, dtype: int64

2. ZERO vs NON-ZERO ROWS
all_core_metrics_zero  False  True   All
oneway_occurrence                       
0                          0   976   976
1                        699   277   976
All                      699  1253  1952

3. TOTAL ACTIVITY BY ONEWAY OCCURRENCE


oneway_occurrence,0,1
srch_fr_e,0.0,35873.000
srch_which_got_e,0.0,35856.000
srch_fr_q,0.0,27389.000
srch_which_got_q,0.0,11380.000
booking,0.0,9888.000
ongng_ride,0.0,579.000
done_ride,0.0,4775.000
rides,0.0,9668.000
cancel_ride,0.0,4311.000
drvr_cancel,0.0,1939.000



4. EXACTLY IDENTICAL ONEWAY PAIRS
Total OneWay pairs: 976
Exactly identical pairs: 0
Different pairs: 976

5. BOTH ONEWAY RECORDS HAVE ACTIVITY?
Ward × vehicle pairs where BOTH OneWay records have activity: 0


,oneway_occurrence,0,1
ward_num,vehicle_type,,



6. SAMPLE DIFFERENT ONEWAY PAIRS


,ward_num,vehicle_type,oneway_occurrence,srch_fr_e,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,ongng_ride,done_ride,rides,cancel_ride,drvr_cancel,rider_cancel,earning,dist
0,b_1,All,0,0,0,0,0,0,0,0,0,0,0,0,0,0.000
8,b_1,All,1,70,70,37,19,0,0,0,0,0,0,0,0,0.000
1,b_1,Auto,0,0,0,0,0,0,0,0,0,0,0,0,0,0.000
9,b_1,Auto,1,0,0,0,0,10,1,7,10,2,1,1,1270,58.450
2,b_1,Bike,0,0,0,0,0,0,0,0,0,0,0,0,0,0.000
10,b_1,Bike,1,0,0,0,0,0,0,0,0,0,0,0,0,0.000
3,b_1,Cab,0,0,0,0,0,0,0,0,0,0,0,0,0,0.000
11,b_1,Cab,1,0,0,0,0,9,0,5,9,4,2,2,1237,47.821
16,b_10,All,0,0,0,0,0,0,0,0,0,0,0,0,0,0.000
24,b_10,All,1,184,184,135,61,0,0,0,0,0,0,0,0,0.000


Occurrence 0:
976 / 976 rows have ZERO core activity

Occurrence 1:
699 rows have activity
277 rows have zero activity

and the aggregated operational activity is in occurrence 1:

Searches                  35,873
Searches getting estimate 35,856
Searches for quote        27,389
Quotes received           11,380
Bookings                   9,888
Completed rides            4,775
Total rides                9,668
Cancellations              4,311
Earnings                1,442,350
Distance                54,341.137

Pairs where BOTH OneWay records have activity = 0

So there is no evidence of two meaningful OneWay business categories.

The first OneWay block is consistently an empty source-generated placeholder/template block, while the second OneWay block contains the actual OneWay ward activity.

This also solves the mathematics:

Original funnel rows = 3904

Remove placeholder OneWay rows = 976

3904 - 976 = 2928 and 244 wards
× 3 real trip categories
× 4 vehicle types
= 2928

so the expected cleaned structure is: 244 wards
×
OneWay / InterCity / Rental
×
All / Auto / Bike / Cab

That does not contradict what we found. Our comparison included all non-key columns, including rate/derived fields where NaN values can compare as unequal and other metadata fields may differ. The operational metrics are what established that occurrence 0 is universally inactive.

#We can now start the actual cleaning

We're going to preserve funnel_df exactly as the raw source representation and create a new cleaned DataFrame called **funnel_clean**.

In [15]:
# ================================================================
# NAMMAFLOW — CLEAN FUNNEL PLACEHOLDER RECORDS
# ================================================================

# Always preserve the original DataFrame
funnel_clean = funnel_df.copy()

# ------------------------------------------------
# Identify the two OneWay occurrences using
# ORIGINAL source row order
# ------------------------------------------------

funnel_clean["oneway_occurrence"] = -1

oneway_mask = (
    funnel_clean["trip_category"] == "OneWay"
)

funnel_clean.loc[
    oneway_mask,
    "oneway_occurrence"
] = (
    funnel_clean.loc[oneway_mask]
    .groupby(["ward_num", "vehicle_type"])
    .cumcount()
    .values
)


# ------------------------------------------------
# Remove occurrence 0:
# confirmed zero-activity placeholder rows
# ------------------------------------------------

placeholder_mask = (
    (funnel_clean["trip_category"] == "OneWay") &
    (funnel_clean["oneway_occurrence"] == 0)
)

print("=" * 75)
print("1. PLACEHOLDER ROWS TO REMOVE")
print("=" * 75)

print("Rows identified:", placeholder_mask.sum())


# Remove placeholders
funnel_clean = (
    funnel_clean.loc[~placeholder_mask]
    .drop(columns="oneway_occurrence")
    .reset_index(drop=True)
)


# ------------------------------------------------
# Validate cleaned dataset
# ------------------------------------------------

print("\n" + "=" * 75)
print("2. SHAPE AFTER CLEANING")
print("=" * 75)

print("Original shape:", funnel_df.shape)
print("Cleaned shape :", funnel_clean.shape)


print("\n" + "=" * 75)
print("3. TRIP CATEGORY COUNTS")
print("=" * 75)

print(
    funnel_clean["trip_category"]
    .value_counts()
)


print("\n" + "=" * 75)
print("4. ROWS PER WARD")
print("=" * 75)

print(
    funnel_clean.groupby("ward_num")
    .size()
    .value_counts()
    .sort_index()
)


print("\n" + "=" * 75)
print("5. CLEANED UNIQUE KEY CHECK")
print("=" * 75)

funnel_key = [
    "ward_num",
    "trip_category",
    "vehicle_type"
]

print(
    "Duplicate key rows:",
    funnel_clean.duplicated(
        subset=funnel_key
    ).sum()
)


print("\n" + "=" * 75)
print("6. EXACT DUPLICATE CHECK")
print("=" * 75)

print(
    "Exact duplicate rows:",
    funnel_clean.duplicated().sum()
)

1. PLACEHOLDER ROWS TO REMOVE
Rows identified: 976

2. SHAPE AFTER CLEANING
Original shape: (3904, 29)
Cleaned shape : (2928, 29)

3. TRIP CATEGORY COUNTS
trip_category
InterCity    976
OneWay       976
Rental       976
Name: count, dtype: int64

4. ROWS PER WARD
12    244
Name: count, dtype: int64

5. CLEANED UNIQUE KEY CHECK
Duplicate key rows: 0

6. EXACT DUPLICATE CHECK
Exact duplicate rows: 0


The structure grain is now confirmed as: **1 funnel row = 1 Ward × 1 Trip Category × 1 Vehicle Type** and mathematically: 244 wards
× 3 trip categories
× 4 vehicle types
= **2928 rows**

That means the funnel structural-cleaning issue is finished. Keep funnel_df as the raw version and use funnel_clean from now on.

The next thing we should tackle is the missing values. We should not simply fill them with 0, because many of the missing fields are rates, and a missing rate may simply mean its denominator was zero.

There is already one strong clue: in original city data, both variant columns had exactly 539 missing values, and earlier we saw exactly 539 rows where trip_category = All. We need to verify whether those fields are simply not applicable to the All category.

In [16]:
#This does not change or impute anything. It only tells us why the values are missing.

# ================================================================
# NAMMAFLOW — MISSING VALUE DIAGNOSTICS
# ================================================================

# Preserve raw city data
city_clean = city_df.copy()


def missing_report(df, name):
    print("\n" + "=" * 75)
    print(f"{name} — MISSING VALUE REPORT")
    print("=" * 75)

    report = pd.DataFrame({
        "missing_count": df.isna().sum(),
        "missing_pct": (df.isna().mean() * 100).round(2)
    })

    report = report[
        report["missing_count"] > 0
    ].sort_values(
        "missing_count",
        ascending=False
    )

    display(report)


# ------------------------------------------------
# 1. Overall missing-value reports
# ------------------------------------------------

missing_report(city_clean, "CITY")
missing_report(funnel_clean, "FUNNEL CLEAN")
missing_report(driver_df, "DRIVER")


# ------------------------------------------------
# 2. Investigate CITY variant-column missingness
# ------------------------------------------------

print("\n" + "=" * 75)
print("CITY — VARIANT COLUMN MISSINGNESS BY TRIP CATEGORY")
print("=" * 75)

variant_cols = [
    "srch_fr_q_variant",
    "srch_which_got_q_variant"
]

for col in variant_cols:

    if col in city_clean.columns:

        print(f"\n--- {col} ---")

        print(
            pd.crosstab(
                city_clean["trip_category"],
                city_clean[col].isna(),
                margins=True
            )
        )


# ------------------------------------------------
# 3. RATE COLUMN ZERO-DENOMINATOR DIAGNOSTIC
# ------------------------------------------------

def rate_null_check(df, rate_col, denominator_candidates):

    if rate_col not in df.columns:
        return

    print("\n" + "-" * 75)
    print(f"RATE: {rate_col}")
    print("-" * 75)

    print("Total missing:", df[rate_col].isna().sum())

    for denom in denominator_candidates:

        if denom not in df.columns:
            continue

        missing = df[rate_col].isna()
        zero = df[denom].fillna(0).eq(0)

        print(f"\nCandidate denominator: {denom}")

        print(
            "Missing rate + denominator = 0 :",
            (missing & zero).sum()
        )

        print(
            "Missing rate + denominator > 0 :",
            (missing & ~zero).sum()
        )

        print(
            "Rate present + denominator = 0 :",
            (~missing & zero).sum()
        )


print("\n" + "=" * 75)
print("CITY — RATE NULL PATTERNS")
print("=" * 75)

rate_null_check(
    city_clean,
    "cnvr_rate",
    ["srch_rqst", "booking", "rides"]
)

rate_null_check(
    city_clean,
    "q_accept_rate",
    ["srch_fr_q", "srch_which_got_q"]
)

rate_null_check(
    city_clean,
    "bkng_cancel_rate",
    ["booking", "cancel_ride"]
)


print("\n" + "=" * 75)
print("FUNNEL CLEAN — RATE NULL PATTERNS")
print("=" * 75)

rate_null_check(
    funnel_clean,
    "srch_to_e_rate",
    ["srch_fr_e", "srch_which_got_e"]
)

rate_null_check(
    funnel_clean,
    "e_to_q_srch_rate",
    ["srch_which_got_e", "srch_fr_q"]
)

rate_null_check(
    funnel_clean,
    "q_accept_rate",
    ["srch_fr_q", "srch_which_got_q"]
)

rate_null_check(
    funnel_clean,
    "q_to_bkng_rate",
    ["srch_which_got_q", "booking"]
)

rate_null_check(
    funnel_clean,
    "cnvr_rate",
    ["srch_fr_e", "booking", "rides"]
)

rate_null_check(
    funnel_clean,
    "bkng_cancel_rate",
    ["booking", "cancel_ride"]
)


CITY — MISSING VALUE REPORT


,missing_count,missing_pct
q_accept_rate,4353,62.66
bkng_cancel_rate,4347,62.57
cnvr_rate,4150,59.74
srch_fr_q_variant,539,7.76
srch_which_got_q_variant,539,7.76



FUNNEL CLEAN — MISSING VALUE REPORT


,missing_count,missing_pct
q_to_bkng_rate,2617,89.38
q_accept_rate,2615,89.31
e_to_q_srch_rate,2614,89.28
srch_to_e_rate,2614,89.28
cnvr_rate,2614,89.28
bkng_cancel_rate,2459,83.98



DRIVER — MISSING VALUE REPORT


,missing_count,missing_pct



CITY — VARIANT COLUMN MISSINGNESS BY TRIP CATEGORY

--- srch_fr_q_variant ---


ValueError: Conflicting name "All" in margins

The missing-value reports that ran before the error are valid. We already know:

CITY
q_accept_rate              4353 missing (62.66%)
bkng_cancel_rate           4347 missing (62.57%)
cnvr_rate                  4150 missing (59.74%)
srch_fr_q_variant           539 missing (7.76%)
srch_which_got_q_variant    539 missing (7.76%)

FUNNEL CLEAN
q_to_bkng_rate             2617 missing (89.38%)
q_accept_rate              2615 missing (89.31%)
e_to_q_srch_rate           2614 missing (89.28%)
srch_to_e_rate             2614 missing (89.28%)
cnvr_rate                  2614 missing (89.28%)
bkng_cancel_rate           2459 missing (83.98%)

DRIVER
No missing values

The high rate-column missingness looks alarming at first, but it may actually be completely legitimate if the denominator is zero. That's what we'll check now.

In [17]:
#This version avoids pd.crosstab(... margins=True) entirely, so the "All" category cannot conflict.

# ================================================================
# NAMMAFLOW — MISSING VALUE DIAGNOSTICS (CONTINUED / FIXED)
# ================================================================

# ------------------------------------------------
# 1. CITY variant-column missingness
# ------------------------------------------------

print("=" * 75)
print("1. CITY — VARIANT COLUMN MISSINGNESS BY TRIP CATEGORY")
print("=" * 75)

variant_cols = [
    "srch_fr_q_variant",
    "srch_which_got_q_variant"
]

for col in variant_cols:

    if col in city_clean.columns:

        temp = (
            city_clean
            .assign(is_missing=city_clean[col].isna())
            .groupby("trip_category")["is_missing"]
            .agg(
                total_rows="size",
                missing_rows="sum"
            )
        )

        temp["present_rows"] = (
            temp["total_rows"] -
            temp["missing_rows"]
        )

        temp["missing_pct"] = (
            temp["missing_rows"] /
            temp["total_rows"] * 100
        ).round(2)

        print(f"\n--- {col} ---")
        display(temp)


# ------------------------------------------------
# Helper function for rate diagnostics
# ------------------------------------------------

def rate_null_check(df, rate_col, denominator_candidates):

    if rate_col not in df.columns:
        return

    print("\n" + "-" * 75)
    print(f"RATE: {rate_col}")
    print("-" * 75)

    missing = df[rate_col].isna()

    print("Total rows   :", len(df))
    print("Missing rate :", missing.sum())
    print("Present rate :", (~missing).sum())

    for denom in denominator_candidates:

        if denom not in df.columns:
            continue

        zero = df[denom].fillna(0).eq(0)

        print(f"\nCandidate denominator → {denom}")

        print(
            "Missing rate & denominator = 0 :",
            (missing & zero).sum()
        )

        print(
            "Missing rate & denominator > 0 :",
            (missing & ~zero).sum()
        )

        print(
            "Rate present & denominator = 0 :",
            (~missing & zero).sum()
        )


# ------------------------------------------------
# 2. CITY rate diagnostics
# ------------------------------------------------

print("\n" + "=" * 75)
print("2. CITY — RATE NULL PATTERNS")
print("=" * 75)

rate_null_check(
    city_clean,
    "cnvr_rate",
    ["srch_rqst", "booking", "rides"]
)

rate_null_check(
    city_clean,
    "q_accept_rate",
    ["srch_fr_q", "srch_which_got_q"]
)

rate_null_check(
    city_clean,
    "bkng_cancel_rate",
    ["booking", "cancel_ride"]
)


# ------------------------------------------------
# 3. FUNNEL rate diagnostics
# ------------------------------------------------

print("\n" + "=" * 75)
print("3. FUNNEL CLEAN — RATE NULL PATTERNS")
print("=" * 75)

rate_null_check(
    funnel_clean,
    "srch_to_e_rate",
    ["srch_fr_e", "srch_which_got_e"]
)

rate_null_check(
    funnel_clean,
    "e_to_q_srch_rate",
    ["srch_which_got_e", "srch_fr_q"]
)

rate_null_check(
    funnel_clean,
    "q_accept_rate",
    ["srch_fr_q", "srch_which_got_q"]
)

rate_null_check(
    funnel_clean,
    "q_to_bkng_rate",
    ["srch_which_got_q", "booking"]
)

rate_null_check(
    funnel_clean,
    "cnvr_rate",
    ["srch_fr_e", "booking", "rides"]
)

rate_null_check(
    funnel_clean,
    "bkng_cancel_rate",
    ["booking", "cancel_ride"]
)

1. CITY — VARIANT COLUMN MISSINGNESS BY TRIP CATEGORY

--- srch_fr_q_variant ---


,total_rows,missing_rows,present_rows,missing_pct
trip_category,,,,
All,539,539,0,100.0
InterCity,1897,0,1897,0.0
OneWay,2909,0,2909,0.0
Rental,1602,0,1602,0.0



--- srch_which_got_q_variant ---


,total_rows,missing_rows,present_rows,missing_pct
trip_category,,,,
All,539,539,0,100.0
InterCity,1897,0,1897,0.0
OneWay,2909,0,2909,0.0
Rental,1602,0,1602,0.0



2. CITY — RATE NULL PATTERNS

---------------------------------------------------------------------------
RATE: cnvr_rate
---------------------------------------------------------------------------
Total rows   : 6947
Missing rate : 4150
Present rate : 2797

Candidate denominator → srch_rqst
Missing rate & denominator = 0 : 4150
Missing rate & denominator > 0 : 0
Rate present & denominator = 0 : 0

Candidate denominator → booking
Missing rate & denominator = 0 : 1941
Missing rate & denominator > 0 : 2209
Rate present & denominator = 0 : 2406

Candidate denominator → rides
Missing rate & denominator = 0 : 2136
Missing rate & denominator > 0 : 2014
Rate present & denominator = 0 : 2797

---------------------------------------------------------------------------
RATE: q_accept_rate
---------------------------------------------------------------------------
Total rows   : 6947
Missing rate : 4353
Present rate : 2594

Candidate denominator → srch_fr_q
Missing rate & denominator = 0 : 4353


The missing values are not random data-quality problems. They are mostly mathematically undefined rates.

For **city_df**, the pattern is exact:
- cnvr_rate is missing 4,150 times, and all 4,150 occur where srch_rqst = 0.

- q_accept_rate is missing 4,353 times, and all occur where srch_fr_q = 0.

- bkng_cancel_rate is missing 4,347 times, and all occur where booking = 0.

- Both *_variant columns are missing exactly 539 times, and every one of those rows has trip_category = "All".

For **funnel_clean** it is equally clean:
srch_to_e_rate   missing when srch_fr_e = 0
e_to_q_srch_rate missing when srch_which_got_e = 0
q_accept_rate    missing when srch_fr_q = 0
q_to_bkng_rate   missing when srch_which_got_q = 0
cnvr_rate        missing when srch_fr_e = 0
bkng_cancel_rate missing when booking = 0

So we should NOT use mean, median, KNN, or zero-fill these rate columns. A rate like booking cancellations / bookings simply has no mathematical value when bookings are zero.

--> Keep structurally undefined rates as NaN. Treat the two variant fields as “not applicable” for trip_category = All.



#Now we move to Bengaluru scope

 NammaFlow is specifically our Bengaluru mobility intelligence project, the next step is to isolate Bengaluru from the city-level file and verify exactly what temporal data we have.

In [18]:
# ================================================================
# NAMMAFLOW — CREATE BENGALURU CITY DATASET
# ================================================================

# Preserve complete multi-city cleaned data
city_all_clean = city_clean.copy()

# ------------------------------------------------
# Filter Bengaluru only
# ------------------------------------------------

city_bengaluru = (
    city_all_clean[
        city_all_clean["city"]
        .astype(str)
        .str.strip()
        .str.lower()
        .eq("bengaluru")
    ]
    .copy()
    .reset_index(drop=True)
)


print("=" * 75)
print("1. BENGALURU DATASET SHAPE")
print("=" * 75)

print("All-city shape :", city_all_clean.shape)
print("Bengaluru shape:", city_bengaluru.shape)


print("\n" + "=" * 75)
print("2. CITY / CITY CODE VALIDATION")
print("=" * 75)

print(
    city_bengaluru[
        ["city", "city_code"]
    ].value_counts()
)


print("\n" + "=" * 75)
print("3. BENGALURU DATES AVAILABLE")
print("=" * 75)

print(
    city_bengaluru["date"]
    .value_counts()
    .sort_index()
)


print("\n" + "=" * 75)
print("4. HOURS AVAILABLE PER DATE")
print("=" * 75)

print(
    city_bengaluru
    .groupby("date")["hour"]
    .nunique()
)


print("\n" + "=" * 75)
print("5. EXACT HOURS BY DATE")
print("=" * 75)

for date, group in city_bengaluru.groupby("date"):

    print(
        f"\n{date.date()} →",
        sorted(group["hour"].unique().tolist())
    )


print("\n" + "=" * 75)
print("6. TRIP CATEGORY DISTRIBUTION")
print("=" * 75)

print(
    city_bengaluru["trip_category"]
    .value_counts()
)


print("\n" + "=" * 75)
print("7. VEHICLE TYPE DISTRIBUTION")
print("=" * 75)

print(
    city_bengaluru["vehicle_type"]
    .value_counts()
)


print("\n" + "=" * 75)
print("8. BOOKING TYPE DISTRIBUTION")
print("=" * 75)

print(
    city_bengaluru["booking_type"]
    .value_counts()
)


print("\n" + "=" * 75)
print("9. BENGALURU UNIQUE KEY VALIDATION")
print("=" * 75)

blr_key = [
    "date",
    "hour",
    "city",
    "trip_category",
    "vehicle_type",
    "booking_type"
]

print(
    "Duplicate key rows:",
    city_bengaluru.duplicated(
        subset=blr_key
    ).sum()
)

print(
    "Exact duplicates:",
    city_bengaluru.duplicated().sum()
)

1. BENGALURU DATASET SHAPE
All-city shape : (6947, 23)
Bengaluru shape: (811, 23)

2. CITY / CITY CODE VALIDATION
city       city_code
Bengaluru  BLR          811
Name: count, dtype: int64

3. BENGALURU DATES AVAILABLE
date
2026-05-03    310
2026-05-09    303
2026-05-10    198
Name: count, dtype: int64

4. HOURS AVAILABLE PER DATE
date
2026-05-03    24
2026-05-09    24
2026-05-10    16
Name: hour, dtype: int64

5. EXACT HOURS BY DATE

2026-05-03 → [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]

2026-05-09 → [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]

2026-05-10 → [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]

6. TRIP CATEGORY DISTRIBUTION
trip_category
OneWay       383
Rental       222
InterCity    206
Name: count, dtype: int64

7. VEHICLE TYPE DISTRIBUTION
vehicle_type
All     306
Cab     306
Auto    199
Name: count, dtype: int64

8. BOOKING TYPE DISTRIBUTION
booking_type
Normal    511
Purple 

For **city_bengaluru**, we have 811 rows × 23 columns, all correctly mapped to Bengaluru / BLR, with zero duplicate keys and zero exact duplicates.

The temporal coverages findings:
2026-05-03 → 24 hours (00:00–23:00) ✅
2026-05-09 → 24 hours (00:00–23:00) ✅
2026-05-10 → 16 hours (00:00–15:00) ⚠️ partial day

Two other important findings:

Trip categories
OneWay      383
Rental      222
InterCity   206

Vehicle types
All         306
Cab         306
Auto        199

Notice Bengaluru has no explicit Bike rows in this city dataset, while the ward funnel dataset does contain Bike.


The next crucial step:

**Q. what does vehicle_type = "All" mean?**

In [20]:
#calculating hourly booking totals
city_bengaluru.groupby("hour")["booking"].sum()

,booking
hour,
0,6216
1,4194
2,1555
3,1380
4,2503
5,4369
6,7498
7,11374
8,17109


In [21]:
# ================================================================
# NAMMAFLOW — VERIFY WHETHER vehicle_type = "All" IS AGGREGATED
# ================================================================

metrics = [
    "srch_rqst",
    "booking",
    "rides",
    "earning"
]

group_cols = [
    "date",
    "hour",
    "trip_category",
    "booking_type"
]

# All-vehicle rows
all_rows = (
    city_bengaluru[
        city_bengaluru["vehicle_type"] == "All"
    ][group_cols + metrics]
    .copy()
)

# Individual vehicle rows
detail_rows = (
    city_bengaluru[
        city_bengaluru["vehicle_type"] != "All"
    ]
    .copy()
)

# Sum Auto + Cab etc. at the same grain
detail_sum = (
    detail_rows
    .groupby(group_cols, as_index=False)[metrics]
    .sum()
)

# Compare
compare = all_rows.merge(
    detail_sum,
    on=group_cols,
    how="left",
    suffixes=("_all", "_detail")
)

print("=" * 70)
print("ALL VEHICLE vs SUM OF DETAIL VEHICLES")
print("=" * 70)

for metric in metrics:

    valid = compare[f"{metric}_detail"].notna()

    matches = (
        compare.loc[valid, f"{metric}_all"]
        ==
        compare.loc[valid, f"{metric}_detail"]
    )

    print(f"\n{metric}")
    print("Rows compared :", valid.sum())
    print("Matches       :", matches.sum())
    print("Different     :", (~matches).sum())

    if valid.sum() > 0:
        print(
            "Match %       :",
            round(matches.mean() * 100, 2)
        )


print("\n" + "=" * 70)
print("SAMPLE COMPARISON")
print("=" * 70)

display(
    compare[
        group_cols +
        [
            "booking_all",
            "booking_detail",
            "rides_all",
            "rides_detail",
            "earning_all",
            "earning_detail"
        ]
    ].head(20)
)

ALL VEHICLE vs SUM OF DETAIL VEHICLES

srch_rqst
Rows compared : 306
Matches       : 0
Different     : 306
Match %       : 0.0

booking
Rows compared : 306
Matches       : 45
Different     : 261
Match %       : 14.71

rides
Rows compared : 306
Matches       : 49
Different     : 257
Match %       : 16.01

earning
Rows compared : 306
Matches       : 81
Different     : 225
Match %       : 26.47

SAMPLE COMPARISON


,date,hour,trip_category,booking_type,booking_all,booking_detail,rides_all,rides_detail,earning_all,earning_detail
0,2026-05-03,0,InterCity,Normal,0,3,0,2,0,3800
1,2026-05-03,0,OneWay,Normal,0,2100,0,1884,0,266239
2,2026-05-03,0,Rental,Normal,0,1,0,1,0,0
3,2026-05-03,1,InterCity,Normal,0,0,0,0,0,0
4,2026-05-03,1,OneWay,Normal,0,1445,0,1308,0,186726
5,2026-05-03,1,Rental,Normal,0,0,0,0,0,0
6,2026-05-03,2,InterCity,Normal,0,1,0,1,0,1830
7,2026-05-03,2,OneWay,Normal,0,532,0,495,0,85798
8,2026-05-03,2,Rental,Normal,0,2,0,2,0,0
9,2026-05-03,3,InterCity,Normal,0,7,0,7,0,0


So we must not treat All as Auto + Cab.

What this likely means structurally is that the All rows are being used for a different stage of the funnel — especially the search/demand side — while vehicle-specific rows carry much of the booking, ride, cancellation and earnings activity.

That would also explain why srch_rqst had 0% exact matches:
overall search requests do not necessarily equal the sum of vehicle-specific search/request figures.

So the correct rule is not “remove All” and also not “sum everything together.”

We need one final semantic profile of All vs Auto/Cab. This will tell us which metrics should come from which row type.

In [22]:
# ================================================================
# NAMMAFLOW — PROFILE 'ALL' VS DETAIL VEHICLE ROWS
# ================================================================

profile_metrics = [
    "srch_rqst",
    "srch_which_got_e",
    "srch_fr_q",
    "srch_which_got_q",
    "booking",
    "rides",
    "done_ride",
    "earning",
    "cancel_ride",
    "drvr_cancel",
    "rider_cancel"
]

profile_metrics = [
    c for c in profile_metrics
    if c in city_bengaluru.columns
]


print("=" * 75)
print("1. ACTIVITY PROFILE — ALL VEHICLE ROWS")
print("=" * 75)

all_profile = []

all_df = city_bengaluru[
    city_bengaluru["vehicle_type"] == "All"
].copy()

for col in profile_metrics:

    all_profile.append({
        "metric": col,
        "rows": len(all_df),
        "non_zero_rows": all_df[col].fillna(0).ne(0).sum(),
        "zero_rows": all_df[col].fillna(0).eq(0).sum(),
        "total": all_df[col].fillna(0).sum()
    })

display(pd.DataFrame(all_profile))


print("\n" + "=" * 75)
print("2. ACTIVITY PROFILE — DETAIL VEHICLE ROWS")
print("=" * 75)

detail_df = city_bengaluru[
    city_bengaluru["vehicle_type"] != "All"
].copy()

detail_profile = []

for col in profile_metrics:

    detail_profile.append({
        "metric": col,
        "rows": len(detail_df),
        "non_zero_rows": detail_df[col].fillna(0).ne(0).sum(),
        "zero_rows": detail_df[col].fillna(0).eq(0).sum(),
        "total": detail_df[col].fillna(0).sum()
    })

display(pd.DataFrame(detail_profile))


print("\n" + "=" * 75)
print("3. SAMPLE ALL ROWS WITH FULL FUNNEL METRICS")
print("=" * 75)

display(
    all_df[
        [
            "date",
            "hour",
            "trip_category",
            "vehicle_type",
            "booking_type"
        ] + profile_metrics
    ].head(20)
)


print("\n" + "=" * 75)
print("4. SAMPLE DETAIL ROWS WITH FULL FUNNEL METRICS")
print("=" * 75)

display(
    detail_df[
        [
            "date",
            "hour",
            "trip_category",
            "vehicle_type",
            "booking_type"
        ] + profile_metrics
    ].head(20)
)

1. ACTIVITY PROFILE — ALL VEHICLE ROWS


,metric,rows,non_zero_rows,zero_rows,total
0,srch_rqst,306,306,0,1151407
1,srch_which_got_e,306,306,0,1146319
2,srch_fr_q,306,306,0,837087
3,srch_which_got_q,306,306,0,464824
4,booking,306,0,306,0
5,rides,306,0,306,0
6,done_ride,306,0,306,0
7,earning,306,0,306,0
8,cancel_ride,306,0,306,0
9,drvr_cancel,306,0,306,0



2. ACTIVITY PROFILE — DETAIL VEHICLE ROWS


,metric,rows,non_zero_rows,zero_rows,total
0,srch_rqst,505,0,505,0
1,srch_which_got_e,505,0,505,0
2,srch_fr_q,505,0,505,0
3,srch_which_got_q,505,0,505,0
4,booking,505,424,81,442586
5,rides,505,416,89,440111
6,done_ride,505,375,130,282542
7,earning,505,375,130,53494253
8,cancel_ride,505,401,104,148641
9,drvr_cancel,505,350,155,74270



3. SAMPLE ALL ROWS WITH FULL FUNNEL METRICS


,date,hour,trip_category,vehicle_type,booking_type,srch_rqst,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,rides,done_ride,earning,cancel_ride,drvr_cancel,rider_cancel
0,2026-05-03,0,InterCity,All,Normal,18,18,18,18,0,0,0,0,0,0,0
2,2026-05-03,0,OneWay,All,Normal,5132,5118,3781,2390,0,0,0,0,0,0,0
5,2026-05-03,0,Rental,All,Normal,11,11,11,11,0,0,0,0,0,0,0
8,2026-05-03,1,InterCity,All,Normal,24,21,18,18,0,0,0,0,0,0,0
10,2026-05-03,1,OneWay,All,Normal,4228,4205,3403,1766,0,0,0,0,0,0,0
13,2026-05-03,1,Rental,All,Normal,5,5,5,5,0,0,0,0,0,0,0
16,2026-05-03,2,InterCity,All,Normal,22,22,22,22,0,0,0,0,0,0,0
18,2026-05-03,2,OneWay,All,Normal,1350,1342,945,721,0,0,0,0,0,0,0
21,2026-05-03,2,Rental,All,Normal,9,9,9,9,0,0,0,0,0,0,0
24,2026-05-03,3,InterCity,All,Normal,23,19,19,19,0,0,0,0,0,0,0



4. SAMPLE DETAIL ROWS WITH FULL FUNNEL METRICS


,date,hour,trip_category,vehicle_type,booking_type,srch_rqst,srch_which_got_e,srch_fr_q,srch_which_got_q,booking,rides,done_ride,earning,cancel_ride,drvr_cancel,rider_cancel
1,2026-05-03,0,InterCity,Cab,Normal,0,0,0,0,3,2,1,3800,2,0,2
3,2026-05-03,0,OneWay,Auto,Normal,0,0,0,0,1164,1164,663,126608,501,264,237
4,2026-05-03,0,OneWay,Cab,Normal,0,0,0,0,936,720,383,139631,524,134,378
6,2026-05-03,0,Rental,Auto,Normal,0,0,0,0,0,0,0,0,0,0,0
7,2026-05-03,0,Rental,Cab,Normal,0,0,0,0,1,1,0,0,1,0,1
9,2026-05-03,1,InterCity,Cab,Normal,0,0,0,0,0,0,0,0,0,0,0
11,2026-05-03,1,OneWay,Auto,Normal,0,0,0,0,785,785,400,83625,385,198,187
12,2026-05-03,1,OneWay,Cab,Normal,0,0,0,0,660,523,240,103101,407,126,271
14,2026-05-03,1,Rental,Auto,Normal,0,0,0,0,0,0,0,0,0,0,0
15,2026-05-03,1,Rental,Cab,Normal,0,0,0,0,0,0,0,0,0,0,0


For NammaFlow, our rule from now onward is:
**Demand / funnel analysis: use vehicle_type == "All".**

**Bookings, completed rides, cancellations, earnings and vehicle analysis: use vehicle_type != "All".**

And we should never blindly sum "All" + Auto + Cab.

That also means the hourly bookings I calculated earlier by grouping the full city_bengaluru table should be ignored. We'll rebuild hourly KPIs correctly from the operational rows.

#We can now close the main Data Understanding phase
Before moving to the next notebook, let's create the clean analytical datasets and flag the partial day.

In [23]:
# ================================================================
# NAMMAFLOW — FINAL ANALYTICAL DATASETS
# ================================================================

# ------------------------------------------------
# 1. DEMAND / SEARCH DATA
# ------------------------------------------------

city_demand = (
    city_bengaluru[
        city_bengaluru["vehicle_type"] == "All"
    ]
    .copy()
    .reset_index(drop=True)
)


# ------------------------------------------------
# 2. VEHICLE / OPERATIONAL DATA
# ------------------------------------------------

city_operations = (
    city_bengaluru[
        city_bengaluru["vehicle_type"] != "All"
    ]
    .copy()
    .reset_index(drop=True)
)


# ------------------------------------------------
# 3. Mark complete vs partial dates
# ------------------------------------------------

hours_per_date = (
    city_bengaluru
    .groupby("date")["hour"]
    .nunique()
)

complete_dates = (
    hours_per_date[
        hours_per_date == 24
    ]
    .index
)

city_demand["is_complete_day"] = (
    city_demand["date"].isin(complete_dates)
)

city_operations["is_complete_day"] = (
    city_operations["date"].isin(complete_dates)
)


# ------------------------------------------------
# 4. Validation
# ------------------------------------------------

print("=" * 75)
print("NAMMAFLOW — FINAL DATA UNDERSTANDING VALIDATION")
print("=" * 75)

print("\nCITY — RAW BENGALURU")
print("Shape:", city_bengaluru.shape)

print("\nCITY — DEMAND DATA")
print("Shape:", city_demand.shape)
print(
    "Vehicle types:",
    city_demand["vehicle_type"].unique()
)

print("\nCITY — OPERATIONS DATA")
print("Shape:", city_operations.shape)
print(
    "Vehicle types:",
    city_operations["vehicle_type"].unique()
)

print("\nWARD FUNNEL CLEAN")
print("Shape:", funnel_clean.shape)

print("\nDRIVER DATA")
print("Shape:", driver_df.shape)


print("\n" + "=" * 75)
print("COMPLETE / PARTIAL DATE STATUS")
print("=" * 75)

date_status = (
    city_bengaluru
    .groupby("date")["hour"]
    .nunique()
    .reset_index(name="hours_available")
)

date_status["day_status"] = (
    date_status["hours_available"]
    .apply(
        lambda x: "Complete"
        if x == 24
        else "Partial"
    )
)

display(date_status)


print("\n" + "=" * 75)
print("FINAL DUPLICATE VALIDATION")
print("=" * 75)

print(
    "Demand exact duplicates:",
    city_demand.duplicated().sum()
)

print(
    "Operations exact duplicates:",
    city_operations.duplicated().sum()
)

print(
    "Funnel exact duplicates:",
    funnel_clean.duplicated().sum()
)

print(
    "Driver exact duplicates:",
    driver_df.duplicated().sum()
)

NAMMAFLOW — FINAL DATA UNDERSTANDING VALIDATION

CITY — RAW BENGALURU
Shape: (811, 23)

CITY — DEMAND DATA
Shape: (306, 24)
Vehicle types: ['All']

CITY — OPERATIONS DATA
Shape: (505, 24)
Vehicle types: ['Cab' 'Auto']

WARD FUNNEL CLEAN
Shape: (2928, 29)

DRIVER DATA
Shape: (486, 6)

COMPLETE / PARTIAL DATE STATUS


,date,hours_available,day_status
0,2026-05-03,24,Complete
1,2026-05-09,24,Complete
2,2026-05-10,16,Partial



FINAL DUPLICATE VALIDATION
Demand exact duplicates: 0
Operations exact duplicates: 0
Funnel exact duplicates: 0
Driver exact duplicates: 0


In [24]:
# ================================================================
# NAMMAFLOW — SAVE PROCESSED DATASETS
# ================================================================

import os

# Create processed-data directory
os.makedirs("data/processed", exist_ok=True)


# ------------------------------------------------
# Save analytical datasets
# ------------------------------------------------

city_demand.to_csv(
    "data/processed/city_demand_bengaluru.csv",
    index=False
)

city_operations.to_csv(
    "data/processed/city_operations_bengaluru.csv",
    index=False
)

funnel_clean.to_csv(
    "data/processed/ward_funnel_clean.csv",
    index=False
)

driver_df.to_csv(
    "data/processed/driver_supply_clean.csv",
    index=False
)

date_status.to_csv(
    "data/processed/date_coverage.csv",
    index=False
)


# ------------------------------------------------
# Validate saved files
# ------------------------------------------------

print("=" * 75)
print("NAMMAFLOW — PROCESSED DATA SAVED")
print("=" * 75)

files = [
    "data/processed/city_demand_bengaluru.csv",
    "data/processed/city_operations_bengaluru.csv",
    "data/processed/ward_funnel_clean.csv",
    "data/processed/driver_supply_clean.csv",
    "data/processed/date_coverage.csv"
]

for file in files:

    size_kb = os.path.getsize(file) / 1024

    print(
        f"✓ {file}"
        f"   ({size_kb:.2f} KB)"
    )


print("\nAll processed datasets saved successfully.")

NAMMAFLOW — PROCESSED DATA SAVED
✓ data/processed/city_demand_bengaluru.csv   (29.47 KB)
✓ data/processed/city_operations_bengaluru.csv   (49.06 KB)
✓ data/processed/ward_funnel_clean.csv   (248.37 KB)
✓ data/processed/driver_supply_clean.csv   (11.36 KB)
✓ data/processed/date_coverage.csv   (0.10 KB)

All processed datasets saved successfully.


So Phase 1 — Data Understanding & Structural Cleaning is complete. We now have the clean inputs needed for EDA:

- city_demand_bengaluru.csv → search/demand funnel

- city_operations_bengaluru.csv → Auto/Cab bookings, rides, cancellations, earnings

- ward_funnel_clean.csv → 244-ward analysis

- driver_supply_clean.csv → ward-level driver supply

- date_coverage.csv → complete vs partial-day metadata

In [25]:
!pwd
!git status

/content/NammaFlow
On branch main
Your branch is up to date with 'origin/main'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	data/processed/

nothing added to commit but untracked files present (use "git add" to track)


In [26]:
!find . -maxdepth 3 -name "01_data_understanding.ipynb" -print
!git status --short
!git ls-files | grep "01_data_understanding"

./notebooks/01_data_understanding.ipynb
?? data/processed/
notebooks/01_data_understanding.ipynb


In [27]:
from google.colab import drive
drive.mount('/content/drive')

!find /content/drive/MyDrive -name "01_data_understanding.ipynb" -print

Mounted at /content/drive
/content/drive/MyDrive/Colab Notebooks/01_data_understanding.ipynb
